File: apps/web/src/lib/firebase.ts
// Firebase Firestore utilities and types
import { 
  collection, 
  doc, 
  getDoc, 
  setDoc, 
  updateDoc, 
  deleteDoc, 
  query, 
  where, 
  orderBy, 
  getDocs,
  Timestamp,
  QueryConstraint
} from "firebase/firestore";
import { db } from "@/firebase.config";
import { sanitizeTextField } from "./security";

// Guard: db is null during SSR/build — these functions are only called from client-side
function getDb() {
  if (!db) throw new Error("Firebase Firestore is not initialized — call from client side only");
  return db;
}

// ============================================================
// TYPES
// ============================================================

export interface Profile {
  id: string; // Firebase UID
  company_name: string;
  main_commodity: string;
  phone_number?: string;
  legal_entity_type: string;
  photo_url?: string;
  google_form_url?: string; // User's own Google Form URL for buyer reviews
  role?: "user" | "admin";
  disabled?: boolean;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface BuyerInquiry {
  id: string;
  user_id: string;
  buyer_name: string;
  country: string;
  product: string;
  quantity: string;
  status: "New" | "Negotiating" | "Closed";
  price_offer?: string;
  notes?: string;
  negotiation_notes?: string;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface ShipmentTracker {
  id: string;
  user_id: string;
  container_number: string;
  shipping_line: string;
  commodity: string;
  status: "Customs" | "On Vessel" | "Arrived";
  etd: Timestamp;
  eta: Timestamp;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface Supplier {
  id: string;
  user_id: string;
  name: string;
  location: string;
  supply_capacity: string;
  created_at: Timestamp;
  updated_at: Timestamp;
}

// ============================================================
// PROFILES
// ============================================================

export async function getProfile(userId: string): Promise<Profile | null> {
  const docRef = doc(getDb(), "profiles", userId);
  const docSnap = await getDoc(docRef);
  return docSnap.exists() ? (docSnap.data() as Profile) : null;
}

export async function createProfile(userId: string, data: Omit<Profile, "id" | "created_at" | "updated_at">) {
  const docRef = doc(getDb(), "profiles", userId);
  await setDoc(docRef, {
    company_name: sanitizeTextField(data.company_name, 150),
    main_commodity: sanitizeTextField(data.main_commodity, 100),
    phone_number: data.phone_number ? sanitizeTextField(data.phone_number, 30) : "",
    legal_entity_type: sanitizeTextField(data.legal_entity_type, 50),
    photo_url: data.photo_url ? sanitizeTextField(data.photo_url, 200000) : "",
    role: data.role || "user",
    disabled: data.disabled || false,
    created_at: Timestamp.now(),
    updated_at: Timestamp.now(),
  });
}

export async function updateProfile(userId: string, data: Partial<Profile>) {
  const docRef = doc(getDb(), "profiles", userId);
  const sanitized: Partial<Profile> = {};
  if (data.company_name !== undefined) sanitized.company_name = sanitizeTextField(data.company_name, 150);
  if (data.main_commodity !== undefined) sanitized.main_commodity = sanitizeTextField(data.main_commodity, 100);
  if (data.phone_number !== undefined) sanitized.phone_number = sanitizeTextField(data.phone_number, 30);
  if (data.legal_entity_type !== undefined) sanitized.legal_entity_type = sanitizeTextField(data.legal_entity_type, 50);
  if (data.photo_url !== undefined) sanitized.photo_url = sanitizeTextField(data.photo_url, 200000);
  if (data.google_form_url !== undefined) sanitized.google_form_url = sanitizeTextField(data.google_form_url, 500);
  if (data.role !== undefined) sanitized.role = data.role;
  if (data.disabled !== undefined) sanitized.disabled = data.disabled;

  await updateDoc(docRef, {
    ...sanitized,
    updated_at: Timestamp.now(),
  });
}

// ============================================================
// BUYER INQUIRIES
// ============================================================

export async function getBuyerInquiries(userId: string): Promise<BuyerInquiry[]> {
  const q = query(
    collection(getDb(), "buyer_inquiries"),
    where("user_id", "==", userId),
    orderBy("created_at", "desc")
  );
  const querySnapshot = await getDocs(q);
  return querySnapshot.docs.map(d => ({ id: d.id, ...d.data() } as BuyerInquiry));
}

export async function createBuyerInquiry(userId: string, data: Omit<BuyerInquiry, "id" | "user_id" | "created_at" | "updated_at">) {
  const docRef = doc(collection(getDb(), "buyer_inquiries"));
  await setDoc(docRef, {
    user_id: userId,
    buyer_name: sanitizeTextField(data.buyer_name, 150),
    country: sanitizeTextField(data.country, 100),
    product: sanitizeTextField(data.product, 150),
    quantity: sanitizeTextField(data.quantity, 50),
    status: data.status,
    price_offer: data.price_offer ? sanitizeTextField(data.price_offer, 50) : "",
    notes: data.notes ? sanitizeTextField(data.notes, 1000) : "",
    negotiation_notes: data.negotiation_notes ? sanitizeTextField(data.negotiation_notes, 1000) : "",
    created_at: Timestamp.now(),
    updated_at: Timestamp.now(),
  });
  return docRef.id;
}

export async function updateBuyerInquiry(inquiryId: string, data: Partial<BuyerInquiry>) {
  const docRef = doc(getDb(), "buyer_inquiries", inquiryId);
  const sanitized: any = {};
  if (data.buyer_name !== undefined) sanitized.buyer_name = sanitizeTextField(data.buyer_name, 150);
  if (data.country !== undefined) sanitized.country = sanitizeTextField(data.country, 100);
  if (data.product !== undefined) sanitized.product = sanitizeTextField(data.product, 150);
  if (data.quantity !== undefined) sanitized.quantity = sanitizeTextField(data.quantity, 50);
  if (data.status !== undefined) sanitized.status = data.status;
  if (data.price_offer !== undefined) sanitized.price_offer = sanitizeTextField(data.price_offer, 50);
  if (data.notes !== undefined) sanitized.notes = sanitizeTextField(data.notes, 1000);
  if (data.negotiation_notes !== undefined) sanitized.negotiation_notes = sanitizeTextField(data.negotiation_notes, 1000);

  await updateDoc(docRef, {
    ...sanitized,
    updated_at: Timestamp.now(),
  });
}

export async function deleteBuyerInquiry(inquiryId: string) {
  const docRef = doc(getDb(), "buyer_inquiries", inquiryId);
  await deleteDoc(docRef);
}

// ============================================================
// SHIPMENT TRACKER
// ============================================================

export async function getShipments(userId: string): Promise<ShipmentTracker[]> {
  const q = query(
    collection(getDb(), "shipment_tracker"),
    where("user_id", "==", userId),
    orderBy("created_at", "desc")
  );
  const querySnapshot = await getDocs(q);
  return querySnapshot.docs.map(d => ({ id: d.id, ...d.data() } as ShipmentTracker));
}

export async function createShipment(userId: string, data: Omit<ShipmentTracker, "id" | "user_id" | "created_at" | "updated_at">) {
  const docRef = doc(collection(getDb(), "shipment_tracker"));
  await setDoc(docRef, {
    user_id: userId,
    container_number: sanitizeTextField(data.container_number, 50),
    shipping_line: sanitizeTextField(data.shipping_line, 100),
    commodity: sanitizeTextField(data.commodity, 100),
    status: data.status,
    etd: data.etd,
    eta: data.eta,
    created_at: Timestamp.now(),
    updated_at: Timestamp.now(),
  });
  return docRef.id;
}

export async function updateShipment(shipmentId: string, data: Partial<ShipmentTracker>) {
  const docRef = doc(getDb(), "shipment_tracker", shipmentId);
  const sanitized: any = {};
  if (data.container_number !== undefined) sanitized.container_number = sanitizeTextField(data.container_number, 50);
  if (data.shipping_line !== undefined) sanitized.shipping_line = sanitizeTextField(data.shipping_line, 100);
  if (data.commodity !== undefined) sanitized.commodity = sanitizeTextField(data.commodity, 100);
  if (data.status !== undefined) sanitized.status = data.status;
  if (data.etd !== undefined) sanitized.etd = data.etd;
  if (data.eta !== undefined) sanitized.eta = data.eta;

  await updateDoc(docRef, {
    ...sanitized,
    updated_at: Timestamp.now(),
  });
}

// ============================================================
// SUPPLIERS
// ============================================================

export async function getSuppliers(userId: string): Promise<Supplier[]> {
  const q = query(
    collection(getDb(), "suppliers"),
    where("user_id", "==", userId),
    orderBy("created_at", "desc")
  );
  const querySnapshot = await getDocs(q);
  return querySnapshot.docs.map(d => ({ id: d.id, ...d.data() } as Supplier));
}

export async function createSupplier(userId: string, data: Omit<Supplier, "id" | "user_id" | "created_at" | "updated_at">) {
  const docRef = doc(collection(getDb(), "suppliers"));
  await setDoc(docRef, {
    user_id: userId,
    name: sanitizeTextField(data.name, 150),
    location: sanitizeTextField(data.location, 200),
    supply_capacity: sanitizeTextField(data.supply_capacity, 100),
    created_at: Timestamp.now(),
    updated_at: Timestamp.now(),
  });
  return docRef.id;
}

export async function updateSupplier(supplierId: string, data: Partial<Supplier>) {
  const docRef = doc(getDb(), "suppliers", supplierId);
  const sanitized: any = {};
  if (data.name !== undefined) sanitized.name = sanitizeTextField(data.name, 150);
  if (data.location !== undefined) sanitized.location = sanitizeTextField(data.location, 200);
  if (data.supply_capacity !== undefined) sanitized.supply_capacity = sanitizeTextField(data.supply_capacity, 100);

  await updateDoc(docRef, {
    ...sanitized,
    updated_at: Timestamp.now(),
  });
}

export async function deleteSupplier(supplierId: string) {
  const docRef = doc(getDb(), "suppliers", supplierId);
  await deleteDoc(docRef);
}

// ============================================================
// ADMIN FUNCTIONS
// ============================================================

export async function getAllProfiles(): Promise<Profile[]> {
  const q = query(collection(getDb(), "profiles"), orderBy("company_name", "asc"));
  const querySnapshot = await getDocs(q);
  return querySnapshot.docs.map(d => ({ id: d.id, ...d.data() } as Profile));
}

export async function updateUserRole(uid: string, role: "user" | "admin") {
  const docRef = doc(getDb(), "profiles", uid);
  await updateDoc(docRef, {
    role,
    updated_at: Timestamp.now(),
  });
}

export async function updateUserDisabled(uid: string, disabled: boolean) {
  const docRef = doc(getDb(), "profiles", uid);
  await updateDoc(docRef, {
    disabled,
    updated_at: Timestamp.now(),
  });
}

export async function getAdminStats(): Promise<{ totalUsers: number; totalInquiries: number; totalShipments: number }> {
  const [profilesSnap, inquiriesSnap, shipmentsSnap] = await Promise.all([
    getDocs(collection(getDb(), "profiles")),
    getDocs(collection(getDb(), "buyer_inquiries")),
    getDocs(collection(getDb(), "shipment_tracker")),
  ]);
  return {
    totalUsers: profilesSnap.size,
    totalInquiries: inquiriesSnap.size,
    totalShipments: shipmentsSnap.size,
  };
}

File: apps/api/firebase.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"cloud.google.com/go/firestore"
	firebase "firebase.google.com/go"
	"firebase.google.com/go/auth"
	"google.golang.org/api/option"
)

// ============================================================
// FIREBASE CLIENT INITIALIZATION
// ============================================================

var (
	fsClient   *firestore.Client
	authClient *auth.Client
	ctx        context.Context
)

// InitFirebase initializes the Firebase Admin SDK and Firestore/Auth clients
func InitFirebase() error {
	ctx = context.Background()

	credentialsPath := os.Getenv("FIREBASE_CREDENTIALS_PATH")
	if credentialsPath == "" {
		credentialsPath = "./firebase-service-account.json"
	}

	// Initialize Firebase App with service account credentials
	opt := option.WithCredentialsFile(credentialsPath)
	app, err := firebase.NewApp(ctx, nil, opt)
	if err != nil {
		return fmt.Errorf("error initializing Firebase app: %v", err)
	}

	// Initialize Firestore client
	client, err := app.Firestore(ctx)
	if err != nil {
		return fmt.Errorf("error initializing Firestore client: %v", err)
	}
	fsClient = client

	// Initialize Auth client
	ac, err := app.Auth(ctx)
	if err != nil {
		return fmt.Errorf("error initializing Firebase auth client: %v", err)
	}
	authClient = ac

	log.Println("✓ Firebase (Auth & Firestore) initialized successfully")
	return nil
}

// CloseFirebase closes the Firestore client connection
func CloseFirebase() error {
	if fsClient != nil {
		return fsClient.Close()
	}
	return nil
}

// ============================================================
// DATA MODELS & STRUCTS FOR FIRESTORE
// ============================================================

type Profile struct {
	ID              string `firestore:"id" json:"id"`
	CompanyName     string `firestore:"company_name" json:"company_name"`
	MainCommodity   string `firestore:"main_commodity" json:"main_commodity"`
	PhoneNumber     string `firestore:"phone_number" json:"phone_number"`
	LegalEntityType string `firestore:"legal_entity_type" json:"legal_entity_type"`
	GoogleFormURL   string `firestore:"google_form_url,omitempty" json:"google_form_url,omitempty"`
	CreatedAt       int64  `firestore:"created_at" json:"created_at"`
	UpdatedAt       int64  `firestore:"updated_at" json:"updated_at"`
}

type FBInquiry struct {
	ID                   string `firestore:"id" json:"id"`
	UserID               string `firestore:"user_id" json:"user_id"`
	BuyerName            string `firestore:"buyer_name" json:"buyerName"`
	Country              string `firestore:"country" json:"country"`
	Product              string `firestore:"product" json:"product"`
	Quantity             string `firestore:"quantity" json:"quantity"`
	Price                string `firestore:"price" json:"price"`
	Status               string `firestore:"status" json:"status"`
	Source               string `firestore:"source" json:"source"`
	Date                 string `firestore:"date" json:"date"`
	LastContactDate      string `firestore:"last_contact_date" json:"lastContactDate"`
	FollowUpReminderDate string `firestore:"follow_up_reminder_date,omitempty" json:"followUpReminderDate,omitempty"`
	NegotiationNotes     string `firestore:"negotiation_notes,omitempty" json:"negotiationNotes,omitempty"`
	FirstEmailSentAt     string `firestore:"first_email_sent_at,omitempty" json:"firstEmailSentAt,omitempty"`
	CreatedAt            int64  `firestore:"created_at" json:"createdAt"`
	UpdatedAt            int64  `firestore:"updated_at" json:"updatedAt"`
}

type FBShipmentDocument struct {
	Type     string `firestore:"type" json:"type"`
	Uploaded bool   `firestore:"uploaded" json:"uploaded"`
}

type FBShipment struct {
	ID              string               `firestore:"id" json:"id"`
	UserID          string               `firestore:"user_id" json:"user_id"`
	ContainerNumber string               `firestore:"container_number" json:"containerNumber"`
	ShippingLine    string               `firestore:"shipping_line" json:"shippingLine"`
	Commodity       string               `firestore:"commodity" json:"commodity"`
	Status          string               `firestore:"status" json:"status"`
	ETD             string               `firestore:"etd" json:"etd"`
	ETA             string               `firestore:"eta" json:"eta"`
	PortOrigin      string               `firestore:"port_origin" json:"portOrigin"`
	PortDestination string               `firestore:"port_destination" json:"portDestination"`
	BuyerName       string               `firestore:"buyer_name" json:"buyerName,omitempty"`
	Documents       []FBShipmentDocument `firestore:"documents" json:"documents"`
	CreatedAt       int64                `firestore:"created_at" json:"createdAt"`
	UpdatedAt       int64                `firestore:"updated_at" json:"updatedAt"`
}

type FBSupplierTransaction struct {
	Date     string `firestore:"date" json:"date"`
	Product  string `firestore:"product" json:"product"`
	Quantity string `firestore:"quantity" json:"quantity"`
	Price    string `firestore:"price" json:"price"`
}

type FBSupplier struct {
	ID                 string                  `firestore:"id" json:"id"`
	UserID             string                  `firestore:"user_id" json:"user_id"`
	Name               string                  `firestore:"name" json:"name"`
	Location           string                  `firestore:"location" json:"location"`
	Product            string                  `firestore:"product" json:"product"`
	SupplyCapacity     string                  `firestore:"supply_capacity" json:"supplyCapacity"`
	LastPrice          string                  `firestore:"last_price" json:"lastPrice"`
	QualityGrade       string                  `firestore:"quality_grade" json:"qualityGrade"`
	ReliabilityScore   int                     `firestore:"reliability_score" json:"reliabilityScore"`
	LegalDocs          bool                    `firestore:"legal_docs" json:"legalDocs"`
	Notes              string                  `firestore:"notes" json:"notes"`
	TransactionHistory []FBSupplierTransaction `firestore:"transaction_history" json:"transactionHistory"`
	CreatedAt          string                  `firestore:"created_at" json:"createdAt"`
	UpdatedAt          int64                   `firestore:"updated_at" json:"updatedAt"`
}

type FBBuyerDeal struct {
	Date      string `firestore:"date" json:"date"`
	Product   string `firestore:"product" json:"product"`
	Quantity  string `firestore:"quantity" json:"quantity"`
	Incoterms string `firestore:"incoterms" json:"incoterms"`
	Value     string `firestore:"value" json:"value"`
}

type FBBuyerCRM struct {
	ID                 string        `firestore:"id" json:"id"`
	UserID             string        `firestore:"user_id" json:"user_id"`
	BuyerName          string        `firestore:"buyer_name" json:"buyerName"`
	Company            string        `firestore:"company" json:"company"`
	Country            string        `firestore:"country" json:"country"`
	DealHistory        []FBBuyerDeal `firestore:"deal_history" json:"dealHistory"`
	TotalVolume        string        `firestore:"total_volume" json:"totalVolume"`
	PaymentHistory     string        `firestore:"payment_history" json:"paymentHistory"`
	PreferredProducts  []string      `firestore:"preferred_products" json:"preferredProducts"`
	CommunicationNotes string        `firestore:"communication_notes" json:"communicationNotes"`
	TrustLevel         int           `firestore:"trust_level" json:"trustLevel"`
	Preferences        string        `firestore:"preferences" json:"preferences"`
	CreatedAt          int64         `firestore:"created_at" json:"createdAt"`
}

type FBNegotiationNote struct {
	ID        string   `firestore:"id" json:"id"`
	UserID    string   `firestore:"user_id" json:"user_id"`
	BuyerName string   `firestore:"buyer_name" json:"buyerName"`
	Date      string   `firestore:"date" json:"date"`
	Category  string   `firestore:"category" json:"category"`
	Content   string   `firestore:"content" json:"content"`
	Tags      []string `firestore:"tags" json:"tags"`
	CreatedAt int64    `firestore:"created_at" json:"createdAt"`
}

// ============================================================
// FIRESTORE OPERATIONS
// ============================================================

// --- Profile ---

func GetProfile(userID string) (*Profile, error) {
	doc, err := fsClient.Collection("profiles").Doc(userID).Get(ctx)
	if err != nil {
		return nil, err
	}
	var profile Profile
	if err := doc.DataTo(&profile); err != nil {
		return nil, err
	}
	return &profile, nil
}

func CreateProfile(userID string, profile *Profile) error {
	profile.ID = userID
	profile.CreatedAt = time.Now().Unix()
	profile.UpdatedAt = time.Now().Unix()
	_, err := fsClient.Collection("profiles").Doc(userID).Set(ctx, profile)
	return err
}

func UpdateProfile(userID string, updates map[string]interface{}) error {
	updates["updated_at"] = time.Now().Unix()
	fbUpdates := make([]firestore.Update, 0, len(updates))
	for k, v := range updates {
		fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
	}
	_, err := fsClient.Collection("profiles").Doc(userID).Update(ctx, fbUpdates)
	return err
}

// --- Buyer Inquiries ---

func GetBuyerInquiries(userID string) ([]FBInquiry, error) {
	iter := fsClient.Collection("buyer_inquiries").Where("user_id", "==", userID).Documents(ctx)
	docs, err := iter.GetAll()
	if err != nil {
		return nil, err
	}
	var inquiries []FBInquiry
	for _, doc := range docs {
		var inquiry FBInquiry
		if err := doc.DataTo(&inquiry); err != nil {
			return nil, err
		}
		inquiry.ID = doc.Ref.ID
		inquiries = append(inquiries, inquiry)
	}
	return inquiries, nil
}

func CreateBuyerInquiry(userID string, inquiry *FBInquiry) (string, error) {
	inquiry.UserID = userID
	inquiry.CreatedAt = time.Now().Unix()
	inquiry.UpdatedAt = time.Now().Unix()
	ref, _, err := fsClient.Collection("buyer_inquiries").Add(ctx, inquiry)
	if err != nil {
		return "", err
	}
	// Save ID inside doc too
	_, _ = ref.Update(ctx, []firestore.Update{{Path: "id", Value: ref.ID}})
	return ref.ID, nil
}

func UpdateBuyerInquiry(inquiryID string, updates map[string]interface{}) error {
	updates["updated_at"] = time.Now().Unix()
	fbUpdates := make([]firestore.Update, 0, len(updates))
	for k, v := range updates {
		fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
	}
	_, err := fsClient.Collection("buyer_inquiries").Doc(inquiryID).Update(ctx, fbUpdates)
	return err
}

func DeleteBuyerInquiry(inquiryID string) error {
	_, err := fsClient.Collection("buyer_inquiries").Doc(inquiryID).Delete(ctx)
	return err
}

// --- Shipments ---

func GetShipments(userID string) ([]FBShipment, error) {
	iter := fsClient.Collection("shipment_tracker").Where("user_id", "==", userID).Documents(ctx)
	docs, err := iter.GetAll()
	if err != nil {
		return nil, err
	}
	var shipments []FBShipment
	for _, doc := range docs {
		var shipment FBShipment
		if err := doc.DataTo(&shipment); err != nil {
			return nil, err
		}
		shipment.ID = doc.Ref.ID
		shipments = append(shipments, shipment)
	}
	return shipments, nil
}

func CreateShipment(userID string, shipment *FBShipment) (string, error) {
	shipment.UserID = userID
	shipment.CreatedAt = time.Now().Unix()
	shipment.UpdatedAt = time.Now().Unix()
	ref, _, err := fsClient.Collection("shipment_tracker").Add(ctx, shipment)
	if err != nil {
		return "", err
	}
	_, _ = ref.Update(ctx, []firestore.Update{{Path: "id", Value: ref.ID}})
	return ref.ID, nil
}

func UpdateShipment(shipmentID string, updates map[string]interface{}) error {
	updates["updated_at"] = time.Now().Unix()
	fbUpdates := make([]firestore.Update, 0, len(updates))
	for k, v := range updates {
		fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
	}
	_, err := fsClient.Collection("shipment_tracker").Doc(shipmentID).Update(ctx, fbUpdates)
	return err
}

// --- Suppliers ---

func GetSuppliers(userID string) ([]FBSupplier, error) {
	iter := fsClient.Collection("suppliers").Where("user_id", "==", userID).Documents(ctx)
	docs, err := iter.GetAll()
	if err != nil {
		return nil, err
	}
	var suppliers []FBSupplier
	for _, doc := range docs {
		var supplier FBSupplier
		if err := doc.DataTo(&supplier); err != nil {
			return nil, err
		}
		supplier.ID = doc.Ref.ID
		suppliers = append(suppliers, supplier)
	}
	return suppliers, nil
}

func CreateSupplier(userID string, supplier *FBSupplier) (string, error) {
	supplier.UserID = userID
	supplier.UpdatedAt = time.Now().Unix()
	ref, _, err := fsClient.Collection("suppliers").Add(ctx, supplier)
	if err != nil {
		return "", err
	}
	_, _ = ref.Update(ctx, []firestore.Update{{Path: "id", Value: ref.ID}})
	return ref.ID, nil
}

func UpdateSupplier(supplierID string, updates map[string]interface{}) error {
	updates["updated_at"] = time.Now().Unix()
	fbUpdates := make([]firestore.Update, 0, len(updates))
	for k, v := range updates {
		fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
	}
	_, err := fsClient.Collection("suppliers").Doc(supplierID).Update(ctx, fbUpdates)
	return err
}

func DeleteSupplier(supplierID string) error {
	_, err := fsClient.Collection("suppliers").Doc(supplierID).Delete(ctx)
	return err
}

// --- Buyer CRM ---

func GetBuyerCRMs(userID string) ([]FBBuyerCRM, error) {
	iter := fsClient.Collection("buyer_crm").Where("user_id", "==", userID).Documents(ctx)
	docs, err := iter.GetAll()
	if err != nil {
		return nil, err
	}
	var crms []FBBuyerCRM
	for _, doc := range docs {
		var crm FBBuyerCRM
		if err := doc.DataTo(&crm); err != nil {
			return nil, err
		}
		crm.ID = doc.Ref.ID
		crms = append(crms, crm)
	}
	return crms, nil
}

func CreateBuyerCRM(userID string, crm *FBBuyerCRM) (string, error) {
	crm.UserID = userID
	crm.CreatedAt = time.Now().Unix()
	ref, _, err := fsClient.Collection("buyer_crm").Add(ctx, crm)
	if err != nil {
		return "", err
	}
	_, _ = ref.Update(ctx, []firestore.Update{{Path: "id", Value: ref.ID}})
	return ref.ID, nil
}

func UpdateBuyerCRM(crmID string, updates map[string]interface{}) error {
	fbUpdates := make([]firestore.Update, 0, len(updates))
	for k, v := range updates {
		fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
	}
	_, err := fsClient.Collection("buyer_crm").Doc(crmID).Update(ctx, fbUpdates)
	return err
}

// --- Negotiation Notes ---

func GetNegotiationNotes(userID string) ([]FBNegotiationNote, error) {
	iter := fsClient.Collection("negotiation_notes").Where("user_id", "==", userID).Documents(ctx)
	docs, err := iter.GetAll()
	if err != nil {
		return nil, err
	}
	var notes []FBNegotiationNote
	for _, doc := range docs {
		var note FBNegotiationNote
		if err := doc.DataTo(&note); err != nil {
			return nil, err
		}
		note.ID = doc.Ref.ID
		notes = append(notes, note)
	}
	return notes, nil
}

func CreateNegotiationNote(userID string, note *FBNegotiationNote) (string, error) {
	note.UserID = userID
	note.CreatedAt = time.Now().Unix()
	ref, _, err := fsClient.Collection("negotiation_notes").Add(ctx, note)
	if err != nil {
		return "", err
	}
	_, _ = ref.Update(ctx, []firestore.Update{{Path: "id", Value: ref.ID}})
	return ref.ID, nil
}

func DeleteNegotiationNote(noteID string) error {
	_, err := fsClient.Collection("negotiation_notes").Doc(noteID).Delete(ctx)
	return err
}

// ============================================================
// SECURE UPDATE/DELETE FUNCTIONS WITH OWNERSHIP VERIFICATION
// ============================================================

// UpdateBuyerInquirySecure verifies the inquiry belongs to the user before updating
func UpdateBuyerInquirySecure(inquiryID string, userID string, updates map[string]interface{}) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("buyer_inquiries").Doc(inquiryID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		updates["updated_at"] = time.Now().Unix()
		fbUpdates := make([]firestore.Update, 0, len(updates))
		for k, v := range updates {
			fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
		}
		return tx.Update(docRef, fbUpdates)
	})
}

// DeleteBuyerInquirySecure verifies the inquiry belongs to the user before deleting
func DeleteBuyerInquirySecure(inquiryID string, userID string) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("buyer_inquiries").Doc(inquiryID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		return tx.Delete(docRef)
	})
}

// UpdateSupplierSecure verifies the supplier belongs to the user before updating
func UpdateSupplierSecure(supplierID string, userID string, updates map[string]interface{}) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("suppliers").Doc(supplierID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		updates["updated_at"] = time.Now().Unix()
		fbUpdates := make([]firestore.Update, 0, len(updates))
		for k, v := range updates {
			fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
		}
		return tx.Update(docRef, fbUpdates)
	})
}

// DeleteSupplierSecure verifies the supplier belongs to the user before deleting
func DeleteSupplierSecure(supplierID string, userID string) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("suppliers").Doc(supplierID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		return tx.Delete(docRef)
	})
}

// UpdateBuyerCRMSecure verifies the CRM record belongs to the user before updating
func UpdateBuyerCRMSecure(crmID string, userID string, updates map[string]interface{}) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("buyer_crm").Doc(crmID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		fbUpdates := make([]firestore.Update, 0, len(updates))
		for k, v := range updates {
			fbUpdates = append(fbUpdates, firestore.Update{Path: k, Value: v})
		}
		return tx.Update(docRef, fbUpdates)
	})
}

// DeleteNegotiationNoteSecure verifies the note belongs to the user before deleting
func DeleteNegotiationNoteSecure(noteID string, userID string) error {
	return fsClient.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
		docRef := fsClient.Collection("negotiation_notes").Doc(noteID)
		doc, err := tx.Get(docRef)
		if err != nil {
			return err
		}

		ownerID, _ := doc.DataAt("user_id")
		if ownerID != userID {
			return fmt.Errorf("permission denied: resource ownership mismatch")
		}

		return tx.Delete(docRef)
	})
}
